src/librbd/cache/BlockGuard.h

```diff
@@ -59,26 +60,59 @@ class BlockGuard {
   };
   typedef std::vector<BlockIOExtent> BlockIOExtents;
 
-  struct BlockIO {
-    // TODO intrusive_list
-
-    BlockIO() {
+  struct C_BlockIORequest : public Context {
+    CephContext *cct;
+    C_BlockIORequest *next_block_request;
+  
+    C_BlockIORequest(CephContext *cct, C_BlockIORequest *next_block_request)
+      : cct(cct), next_block_request(next_block_request) {
     }
-    BlockIO(uint64_t block, BlockIOExtents &&extents)
-      : block(block), extents(extents) {
+  
+    virtual void finish(int r) override {
+      //ldout(cct, 20) << "(" << get_name() << "): r=" << r << dendl;
+  
+      if (r < 0) {
+        // abort the chain of requests upon failure
```

Comment:
How do you denote the case where the current parent image cache does not have the requested extent and you need to (1) ask the OSDs for the data or (2) re-direct the read request to the next image in the clone hierarchy?